### PR TITLE
Feature: Fixed size list nested type (ARRAY)

### PR DESCRIPTION
**Pull request by Maxxen**
## ARRAY Physical Type

This PR adds a new nested physical+logical type optimized for representing fixed-size-lists, with accompanying functions and full support for storage/joins/grouping/sorting. Internally an `ARRAY` vector is somewhat like a combination of a `STRUCT` and a `LIST`. The vector itself does not store any data except for a validity mask (like structs), but has a single child vector (like lists). 

Unlike lists however there is no need to keep and store "list entries" with offsets and lengths because the data of the array at position `i` will always be at `array_size * i`. Therefore the child vector is always a flat vector and allocated at a fixed `STANDARD_VECTOR_SIZE * array_size` capacity, and is never shrunk or resized (unless the parent vector is too). This makes array vectors less memory efficient when they contain a lot of nulls or not enough elements to fill an entire vector, but in exchange they make child element access faster, but most importantly, predictable, which could enable for more optimized functions to be implemented over them (e.g. list lambdas, sorting, filtering, e.t.c.).

All the list functions should work with arrays, but the following array-native functions are implemented

- `array_value` - Create an ARRAY containing the argument values
- `array_cross_product` - Compute the cross product of two arrays with length 3.
- `array_cosine_similarity` - Compute the cosine similarity between two arrays.
- `array_distance` - Compute the distance between two arrays.
- `array_inner_product` - Compute the inner product between two arrays.
- `length` (overloaded) - Returns the length of the array (always constant).
- `array_length` - Returns the length of the array along a given dimension. (useful for nested arrays).

### `ARRAY` constructor and untangle `array_` as alias for list
We've basically inherited the concept of "arrays" from Postgres and currently have a lot of aliases for list functions (and tests!) using the `array_` prefix. We should probably remove/rename these to make a clear separation between lists and this new array type.

We also have the `ARRAY[]` syntax which I think we also should change to basically alias `array_value` to create this new type instead of lists. As of now the "cleanest" way to create an array literal is to do a cast, e.g. `[1,2,3]::INT[3]`, but that is of course less efficient.

### Specialize list functions for arrays

I've made it so that `LIST` and `ARRAY` are implicitly castable to each other if their child elements are implicitly castable, which makes it so that we can use arrays with all the existing list functions (although by introducing a manual cast in their bind hooks, read on below). This _works_, but is pretty inefficient, not just because we have to allocate a bunch of list_entry_t's but also because I think a lot of list functions could be implemented a lot easier and efficiently when you know all the list have the same length, which is always the case for arrays. This could be done incrementally over time though, so no big deal.

### Example

```sql
-- Construct with the 'array_value' function
SELECT array_value(1,2,3);
┌──────────────────────┐
│ array_value(1, 2, 3) │
│      integer[3]      │
├──────────────────────┤
│ [1, 2, 3]            │
└──────────────────────┘

-- You can always implicitly cast to a list (and use list functions, like list_extract, '[i]')
SELECT array_value(1,2,3)[2];
┌─────────────────────────┐
│ array_value(1, 2, 3)[2] │
│          int32          │
├─────────────────────────┤
│                       2 │
└─────────────────────────┘

-- You can cast from list, but the dimensions have to match up!
SELECT [3,2,1]::INT[3];
┌──────────────────────────────────────────────┐
│ CAST(main.list_value(3, 2, 1) AS INTEGER[3]) │
│                  integer[3]                  │
├──────────────────────────────────────────────┤
│ [3, 2, 1]                                    │
└──────────────────────────────────────────────┘

SELECT [3,2,1]::INT[4];
Error: Conversion Error: Cannot cast list with length 3 to array with length 4

-- Arrays can of course also be nested!
SELECT array_value(array_value(1,2), array_value(3,4), array_value(5,6));
┌──────────────────────────────────────────────────────────────────────┐
│ array_value(array_value(1, 2), array_value(3, 4), array_value(5, 6)) │
│                            integer[2][3]                             │
├──────────────────────────────────────────────────────────────────────┤
│ [[1, 2], [3, 4], [5, 6]]                                             │
└──────────────────────────────────────────────────────────────────────┘

SELECT array_value({'a': 1, 'b': 2}, {'a': 3, 'b': 4});
┌─────────────────────────────────────────────────────────────────────────────────┐
│ array_value(main.struct_pack(a := 1, b := 2), main.struct_pack(a := 3, b := 4)) │
│                         struct(a integer, b integer)[2]                         │
├─────────────────────────────────────────────────────────────────────────────────┤
│ [{'a': 1, 'b': 2}, {'a': 3, 'b': 4}]                                            │
└─────────────────────────────────────────────────────────────────────────────────┘

```

### Generic binding and casts

It is currently a bit awkward to write functions for the list type as you usually want them to be somewhat polymorphic over the child type and do a lot of binding logic and casting in explicit bind functions when implementing e.g. a scalar. For arrays this is even more problematic since you usually also want your function to accept arrays - not just of any type - but also any _size_. So far I've implemented the new functions for the array type as taking `LogicalType::ANY` and then performing all type checking manually in the bind hook. While `LogicalType::ANY` works to signal placeholder types we would need something similar to signal "any number", basically like a [non-type template parameter](https://en.cppreference.com/w/cpp/language/template_parameters).

I think this could be an opportunity to look into extending our type system to handle both the nested ANY type better and perhaps also support non-type parameter placeholders/generic parameters. 

Note that I think this is primarily useful for docs/error messages, I don't mind too much if you have to actually enforce the types manually when binding (similar to how ANY works now I guess).

### Small array optimizations

An interesting idea is to allow "small" arrays of fixed-size types, e.g. `INTEGER[4]`, `DOUBLE[2]` to be stored in the parent vectors data as some sort of small blob/struct up to 16/32 bytes, avoiding having a child vector altogether. You would also probably need to store an additional validity mask for the array elements as an auxiliary vector buffer though.  But since this would change so much of the execution/storage maybe it would make more sense to have this as another separate physical type? (like a `TUPLE` or something). idk.

### Add ARRAY to `all_types`, client libraries/integrations

Notably this type maps better to e.g. Arrow FixedSizeLists and other integrations like numpy arrays.

### CODE REVIEW

**Comment by Alex-Monahan** [comment]
Super exciting! This opens up a lot of fun possibilities. Do you happen to have a list of the future capabilities this could help power? That would be cool to include in the next version release blog.

One other question is around Postgres compatibility. Is there a chance this breaks a number of queries in the wild? Would you be open to a less concise name like `fixed_array` or similar? Or maybe there is a pragma that can be set for higher postgres compatibility, but not loading this new type? (I don't want to have to force a check in every function for "was the no_fixed_arrays pragma set?" One of the selling points of MotherDuck is easy migration / code reuse from Postgres, so it is a big value for us!

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Hi @Maxxen, awesome PR! I will review the code later, but I wanted to do some testing first to see the shiny new feature :)

While testing, I found a few issues. The first few queries all work, but there are some issues with the queries after `mode skip` in the snippet below. Some of the issues we can probably ignore, but there are a few `INTERNAL EXCEPTION` that we can't ignore.

I hope this helps!
```
require json

require parquet

statement ok
PRAGMA verify_external

statement ok
CREATE TABLE arrays (a INTEGER[3])

statement ok
INSERT INTO arrays VALUES ([1, 2, 3]), ([1, 2, 4]), ([7, 8, 9]), ([-1, -2, -3]), (NULL), ([4, NULL, 2])

query I
SELECT a[3] FROM arrays
----
3
4
9
-3
NULL
2

query I
SELECT DISTINCT a FROM arrays ORDER BY ALL
----
[-1, -2, -3]
[1, 2, 3]
[1, 2, 4]
[4, NULL, 2]
[7, 8, 9]
NULL

query I
SELECT DISTINCT a FROM arrays WHERE a[1] > 0 ORDER BY ALL
----
[1, 2, 3]
[1, 2, 4]
[4, NULL, 2]
[7, 8, 9]

query I
SELECT * FROM (
    SELECT a FROM ARRAYS
    UNION
    SELECT a FROM ARRAYS
)
ORDER BY ALL
----
[-1, -2, -3]
[1, 2, 3]
[1, 2, 4]
[4, NULL, 2]
[7, 8, 9]
NULL

query I
SELECT * FROM (
    SELECT a FROM ARRAYS WHERE a[1] > 0
    UNION
    SELECT a FROM ARRAYS WHERE a[1] > 0
)
ORDER BY ALL
----
[1, 2, 3]
[1, 2, 4]
[4, NULL, 2]
[7, 8, 9]

query I
SELECT first(DISTINCT a ORDER BY a) FROM arrays
----
[-1, -2, -3]

query I
SELECT a::VARCHAR FROM arrays ORDER BY ALL
----
[-1, -2, -3]
[1, 2, 3]
[1, 2, 4]
[4, NULL, 2]
[7, 8, 9]
NULL

query I
SELECT TRY_CAST(a::INTEGER[] AS INTEGER[3]) FROM ARRAYS ORDER BY ALL
----
[-1, -2, -3]
[1, 2, 3]
[1, 2, 4]
[4, NULL, 2]
[7, 8, 9]
NULL

query I
SELECT a[2:-1] FROM arrays
----
[2, 3]
[2, 4]
[8, 9]
[-2, -3]
NULL
[NULL, 2]

query I
SELECT a[3:99] FROM arrays
----
[3]
[4]
[9]
[-3]
NULL
[2]

query IIIIII
DESCRIBE SELECT * FROM arrays
----
a	INTEGER[3]	YES	NULL	NULL	NULL

query I
SELECT a.filter(x -> x > 0) FROM arrays
----
[1, 2, 3]
[1, 2, 4]
[7, 8, 9]
[]
NULL
[4, 2]

query I
SELECT a.apply(x -> x + 1)::INTEGER[3] FROM arrays
----
[2, 3, 4]
[2, 3, 5]
[8, 9, 10]
[0, -1, -2]
NULL
[5, NULL, 3]

mode skip

query I
SELECT list(a ORDER BY a[3]) FROM arrays
----
# Shouldn't Internal Exception:
# ABORT THROWN BY INTERNAL EXCEPTION: LIST aggregate not yet implemented for INTEGER[3]

query I
SELECT * FROM test_vector_types(NULL::INTEGER[3])
----
# This is probably OK?
# Error: Not implemented Error: Unimplemented type for test_vector_types INTEGER[3]

query I
SELECT TRY_CAST(test_vector AS INT[2]) AS a FROM test_vector_types(NULL::INTEGER[])
----
# Should not fail but create NULL instead:
# Conversion Error: Cannot cast list with length 0 to array with length 2

# Also, why can't we cast length 0 to 2? Can't we just do list_resize(arr, 2), and then cast?
# This would fill in the gaps with NULL 

query I
SELECT a::JSON FROM arrays
----
# Shouldn't Internal Exception:
# ABORT THROWN BY INTERNAL EXCEPTION: Unsupported type arrived at JSON create function

statement ok
COPY arrays to '__TEST_DIR__/arrays.parquet'
# This is probably OK for now
# Not implemented Error: Unimplemented type for Parquet "INTEGER[3]"

foreach format parquet json csv

statement ok
CREATE OR REPLACE TABLE arrays2 (a INTEGER[3])

statement ok
COPY (SELECT [range, range + 1, range + 2] a FROM range(3)) TO '__TEST_DIR__/arrays.${format}'

statement ok
COPY arrays2 FROM '__TEST_DIR__/arrays.${format}'
# CSV and JSON parser give an error, but probably OK for now
# Invalid Input Error: Unimplemented type for cast (VARCHAR -> INTEGER[3]) at line 1 in column "a". Parser options:
# Not implemented Error: Cannot read a value of type INTEGER[3] from a json file

query I
SELECT * FROM arrays2
----
[0, 1, 2]
[1, 2, 3]
[2, 3, 4]

endloop

query I
SELECT array_value(arrays.*) FROM arrays
----
# This is broken with list_value apparently, if you do this you get empty lists
# If we do this with array_value, we get this error
# Invalid Input Error: array_value requires at least one argument
# Is that OK? I guess we don't ever want to make size 0 arrays, e.g., INTEGER[0]

query I
SELECT '[1, 2, 3]'::INTEGER[3]
----
# Can't we just cast to INTEGER[] and then cast that to INTEGER[3] to make this cast possible?
# Error: Conversion Error: Unimplemented type for cast (VARCHAR -> INTEGER[3])

query I
SELECT a1.a || a2.a FROM arrays a1, arrays a2;
----
# The || operator is shorthand for list_concat, which triggers this error
# ABORT THROWN BY INTERNAL EXCEPTION: Assertion triggered in file "/Users/laurens/git/duckdb/main-duckdb/src/function/scalar/list/list_concat.cpp" on line 90: lhs.id() == LogicalTypeId::LIST

query I
SELECT list_resize(a, 3) FROM arrays
----
# ABORT THROWN BY INTERNAL EXCEPTION: Invalid PhysicalType for GetTypeIdSize

query I
SELECT list_position(a, 1) FROM arrays
----
# ABORT THROWN BY INTERNAL EXCEPTION: Assertion triggered in file "/Users/laurens/git/duckdb/main-duckdb/src/common/types.cpp" on line 871: type.id() == LogicalTypeId::LIST || type.id() == LogicalTypeId::MAP

query I
SELECT list_contains(a, 1) FROM arrays
----
# ABORT THROWN BY INTERNAL EXCEPTION: Assertion triggered in file "/Users/laurens/git/duckdb/main-duckdb/src/common/types.cpp" on line 871: type.id() == LogicalTypeId::LIST || type.id() == LogicalTypeId::MAP

query IIIIIIIIIIII
SUMMARIZE arrays
----
a	INTEGER[3]	[-1, -2, -3]	[7, 8, 9]	5	NULL	NULL	NULL	NULL	NULL	6	16.67%
# wrong result for SUMMARIZE! the minimum value of column a should be [-1, -2, -3], but it returns [1, 2, 3]

# if we investigate this, we see that ORDER BY works fine
# these two queries should return the same
query I
SELECT a FROM arrays ORDER BY a LIMIT 1
----
[-1, -2, -3]

query I
SELECT min(a) FROM arrays
----
[-1, -2, -3]
# wrong result! this one returns [1, 2, 3] instead of [-1, -2, -3]

# the max() value also (sometimes) returns wrong results
# if we insert these values into 'arrays'
statement ok
INSERT INTO arrays VALUES ([-7, -8, -9]), ([-8, -9, -10])

query I
SELECT max(a) FROM arrays
----
[7, 8, 9]
# wrong result! we get back [-8, -9, -10]
```

**Comment by alippai** [comment]
Maybe a fast column wise and row wise, element by element sum (or product) of arrays would add value as well? [1,2,3]+[4,5,6]=[5,7,9]. BLAS functions could be an overkill, but a few fast paths for element wide array operations would make sense.

**Comment by Maxxen** [comment]
Hey @lnkuiper thanks for the thorough testing! Looks like there still are some list function cases where we don't insert proper casts, and probably some min/max comparators I got wrong. I know there are still some ergonomic casts and such missing (like upcasting smaller lists by null-padding), but Id like to focus more on that in a followup PR. I'll add your test cases though and have a look at it next week! No rush in reviewing the code though, this PR is probably too big for everyone to review everything individually, but it would be great if you could just check the execution-layer stuff that I asked you for help on, maybe @taniabogatsch can do the list/cast/binding/frontend-y stuff and @Mytherin have a look at the storage-layer/column collection parts when y'all have time (again, no rush).

@alippai Yes this is something I want to add as well in a followup PR, as it stands right now it is a little tedious to implement generic array functions as you have to do all the type checking manually. But will look into adding more math-related functions in the future for sure!

File: src/common/row_operations/row_heap_gather.cpp [hunk 1]
```diff
@@ -133,6 +133,84 @@ static void HeapGatherListVector(Vector &v, const idx_t vcount, const SelectionV
...
+			auto &child_validity = FlatVector::Validity(child_vector);
+			for (idx_t elem_idx = 0; elem_idx < chunk_size; elem_idx++) {
+				child_validity.Set(array_start + elem_idx, *(array_validitymask_location) & (1 << offset_in_byte));
+				if (++offset_in_byte == 8) {
+					array_validitymask_location++;
+					offset_in_byte = 0;
+				}
+			}
+
+			SelectionVector array_sel(chunk_size);
```
Comment (lnkuiper) [inline src/common/row_operations/row_heap_gather.cpp:187]: Can we initialize the selection vector once outside the loop with `STANDARD_VECTOR_SIZE`? Should always be enough.

File: src/common/row_operations/row_match.cpp [hunk 2]
```diff
@@ -230,6 +230,29 @@ static void TemplatedMatchList(Vector &col, Vector &rows, SelectionVector &sel,
 	}
 }
 
+template <class OP, bool NO_MATCH_SEL>
+static void TemplatedMatchArray(Vector &col, Vector &rows, SelectionVector &sel, idx_t &count,
```
Comment (lnkuiper) [inline src/common/row_operations/row_match.cpp:234]: I just reworked this file in #8979 but resolving the merge conflicts should be easy. You may want to add a test to `test/sql/join/inner/test_join_is_distinct.test` to double-check that everything works for `ARRAY` too.

File: src/common/types/row/tuple_data_scatter_gather.cpp [hunk 3]
```diff
@@ -144,48 +145,69 @@ void TupleDataCollection::ComputeHeapSizes(Vector &heap_sizes_v, const Vector &s
...
 	default:
 		throw NotImplementedException("ComputeHeapSizes for %s", EnumUtil::ToString(source_v.GetType().id()));
 	}
 }
 
-void TupleDataCollection::WithinListHeapComputeSizes(Vector &heap_sizes_v, const Vector &source_v,
-                                                     TupleDataVectorFormat &source_format,
-                                                     const SelectionVector &append_sel, const idx_t append_count,
-                                                     const UnifiedVectorFormat &list_data) {
+void TupleDataCollection::WithinCollectionHeapComputeSizes(Vector &heap_sizes_v, const Vector &source_v,
```
Comment (lnkuiper) [inline src/common/types/row/tuple_data_scatter_gather.cpp:174]: Small nitpick, can we rename this `WithinCollectionComputeHeapSizes`?

File: src/common/types/row/tuple_data_scatter_gather.cpp [hunk 4]
```diff
@@ -1012,12 +1160,75 @@ static void TupleDataListGather(const TupleDataLayout &layout, Vector &row_locat
...
+			target_list_offset += list_length;
+		} else {
+			source_heap_validity.SetInvalid(source_idx);
+			target_validity.SetInvalid(target_idx);
+		}
+	}
+
+	auto list_size_before = 0; // ListVector::GetListSize(target);
+	// ListVector::Reserve(target, list_size_before + target_list_offset);
+	// ListVector::SetListSize(target, list_size_before + target_list_offset);
```
Comment (lnkuiper) [inline src/common/types/row/tuple_data_scatter_gather.cpp:1214]: Can these comments be removed?

File: src/common/vector_operations/vector_hash.cpp [hunk 5]
```diff
@@ -176,6 +176,33 @@ static inline void ListLoopHash(Vector &input, Vector &hashes, const SelectionVe
...
+	input.ToUnifiedFormat(count, idata);
+
+	// Hash the children into a temporary
+	auto &child = ArrayVector::GetEntry(input);
+	auto array_size = ArrayType::GetSize(input.GetType());
+	auto child_count = array_size * count;
+
+	Vector child_hashes(LogicalType::HASH, child_count);
+	if (child_count > 0) {
+		VectorOperations::Hash(child, child_hashes, child_count);
```
Comment (lnkuiper) [inline src/common/vector_operations/vector_hash.cpp:193]: We probably need to add this here
```c++
child_hashes.Flatten(child_count);
```
I recently added this to `ListLoopHash` in #8979 to fix a bug with joining lists. The problem was that the child of the list became `CONSTANT` by the `VectorOperations::Hash`, but in the rest of the function it was assumed that it was `FLAT`.

**Review by lnkuiper — CHANGES_REQUESTED** [review]
I just looked through the code (at least the part you asked me to), and it looks great!

I have some minor comments below, and I would like to ask for a test where a large list of strings goes into a `ColumnDataCollection`, for example:
```sql
CREATE TABLE test AS
SELECT range % 4 g,
       10000 - range o,
       [range, range + 1, range + 2]::VARCHAR[3] a,
       range i
FROM range(10000);
SELECT g o, a, sum(i) OVER (PARTITION BY g ORDER BY o) FROM test;
```
This should force the large array of strings into a `ColumnDataCollection`, and we can verify that the string swizzling works properly.

**Comment by Maxxen** [comment]
Alright, I've fixed all the issues @lnkuiper brought up and added his examples to the tests. Other work:
- You can now roundtrip parquet/json/csv containing ARRAY types, although not optimized (I think parquet performs VARCAHR conversions for now)
- For binding purposes its possible to construct a array type of "unknown size", similar to LogicalType::ANY
- implicit list->array casts are allowed if the target array type is of known size.
- Squashed a whole bunch of bugs regarding child-vector validity at all levels of the system. Similar to structs the validity of the child vector have to be synchronized with the parent (array) vector.

After working on all this (the validity bug fixing in particular) Im actually pretty confident that everything works as it should now.

**Review by taniabogatsch — CHANGES_REQUESTED** [review]
Hi @Maxxen! Really cool PR! I looked at the first part of my review assignment haha `maybe @taniabogatsch can do the list/cast/binding/frontend-y`. I added comments, some nitpicks, and questions!

I'll add another review to cover binding and casts.

### General comments

> We should probably remove/rename these to make a clear separation between lists and this new array type.

I completely agree. As we've already discussed, some list functions even match the new `ARRAY` type better! Let's pick this up in a future PR and clean up our functions there.

> An interesting idea is to allow "small" arrays of fixed-size types [...] to be stored in the parent vectors data [...]

I don't think that this makes a lot of impact performance-wise. We don't assume people to have a few/a single row in their table with a small array size. Then, the child vector will be filled sufficiently for each processed data chunk, and our scalar function execution will carry the performance.

### Other thoughts

> This makes array vectors less memory efficient when they contain a lot of nulls or not enough elements to fill an entire vector, [...]

This comment directly made me wonder how much that impacts aggregation performance, as I fixed this for `LIST` by writing the custom `list_segment` implementation. But then I noticed we do not support `ARRAY` as an aggregate function yet. Would that even make sense to have in the future, as that would require all groups to have the same size?
```sql
D SELECT LIST(i) FROM t;
┌─────────────────────┐
│       list(i)       │
│      int32[][]      │
├─────────────────────┤
│ [[1, 2], [1, 2, 3]] │
└─────────────────────┘
D SELECT ARRAY(i) FROM t;
Error: Parser Error: syntax error at or near "i"
LINE 1: SELECT ARRAY(i) FROM t;
```

File: src/common/extra_type_info.cpp [hunk 6]
```diff
@@ -312,4 +312,17 @@ void EnumTypeInfo::Serialize(Serializer &serializer) const {
 	                     [&](Serializer::List &list, idx_t i) { list.WriteElement(strings[i]); });
 }
 
+//===--------------------------------------------------------------------===//
+// ArrayTypeInfo
+//===--------------------------------------------------------------------===//
+
+ArrayTypeInfo::ArrayTypeInfo(LogicalType child_type_p, uint32_t size_p)
```
[the remark this answers is not in the document]
Comment (Maxxen) [reply src/common/extra_type_info.cpp:319]: This stems from the Postgres parser not supporting larger sizes within the `type[ ]` brackets, but I could definitely bump this to make it consistent with the lists.

File: src/common/types.cpp [hunk 7]
```diff
@@ -381,6 +391,17 @@ string LogicalType::ToString() const {
 		ret += ")";
 		return ret;
 	}
+	case LogicalTypeId::ARRAY: {
+		if (!type_info_) {
+			return "ARRAY";
+		}
+		auto size = ArrayType::GetSize(*this);
+		if (size == 0) {
+			return ArrayType::GetChildType(*this).ToString() + "[ANY]";
```
[the remark this answers is not in the document]
Comment (Maxxen) [reply src/common/types.cpp:400]: During binding failures (and for function docs, e.g. in `duckdb_functions()`) we print the types of the registered function arguments, so this is mostly for documentation purposes.

File: src/include/duckdb/common/types/vector.hpp [hunk 8]
```diff
@@ -447,6 +449,20 @@ struct StructVector {
...
+	DUCKDB_API static const Vector &GetEntry(const Vector &vector);
+	//! Gets a reference to the underlying child-vector of an array
+	DUCKDB_API static Vector &GetEntry(Vector &vector);
+	//! Gets the total size of the underlying child-vector of an array
+	DUCKDB_API static idx_t GetTotalSize(const Vector &vector);
+	//! Allocate fake list entries for a vector
+	//! Note that there is nothing ensuring that the allocated data
+	//! remains valid (e.g. if this vector is resized)
+	//! This is only used during row serialization
+	DUCKDB_API static void AllocateFakeListEntries(Vector &vector);
```
[the remark this answers is not in the document]
Comment (Maxxen) [reply src/include/duckdb/common/types/vector.hpp:463]: Well, you need to have access to the vector internals, so I think this is potentially cleaner than giving the TupleDataCollection friend-access to the vector class.

File: src/include/duckdb/common/types.hpp [hunk 9]
```diff
@@ -424,6 +430,13 @@ struct UnionType {
 	DUCKDB_API static const child_list_t<LogicalType> CopyMemberTypes(const LogicalType &type);
 };
 
+struct ArrayType {
+	DUCKDB_API static const LogicalType &GetChildType(const LogicalType &type);
+	DUCKDB_API static uint32_t GetSize(const LogicalType &type);
+	DUCKDB_API static constexpr uint64_t MAX_ARRAY_SIZE = 32768;
```
[the remark this answers is not in the document]
Comment (Maxxen) [reply src/include/duckdb/common/types.hpp:436]: Will bump it to idx_t!

File: src/core_functions/scalar/array/functions.json [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        "name": "array_cross_product",
+        "parameters": "array, array",
+        "description": "Compute the cross product of two arrays with length 3.",
+        "example": "array_cross_product([1, 2, 3], [1, 2, 3])",
+        "type": "scalar_function_set"
+    },
+    {
+        "name": "array_cosine_similarity",
+        "parameters": "array1,array2",
+        "description": "Compute the cosine similarity between two arrays.",
```
[the remark this answers is not in the document]
Comment (Maxxen) [reply src/core_functions/scalar/array/functions.json:19]: Where else would they go?

File: src/core_functions/function_list.cpp [hunk 11]
```diff
@@ -17,6 +17,7 @@
 #include "duckdb/core_functions/scalar/string_functions.hpp"
 #include "duckdb/core_functions/scalar/struct_functions.hpp"
 #include "duckdb/core_functions/scalar/union_functions.hpp"
+#include "duckdb/core_functions/scalar/array_functions.hpp"
```
[the remark this answers is not in the document]
Comment (Maxxen) [reply src/core_functions/function_list.cpp:20]: idk, where else would they go? I thought everything in main duckdb is core_functions?

[the remark this answers is not in the document]
Comment (taniabogatsch) [reply src/core_functions/function_list.cpp:20]: Yea, tbh, I am also confused about this. I just saw that we also had `src/function/scalar/...` when writing this comment and thought that they belonged there. But how/why do we have these two separate places?

File: src/include/duckdb/common/types/vector.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (taniabogatsch) [reply src/include/duckdb/common/types/vector.hpp:463]: Ah, that makes sense. Thanks for the clarification!

File: .github/config/uncovered_files.csv [hunk 12]
```diff
@@ -58,7 +58,7 @@ common/serializer/buffered_file_reader.cpp	11
 common/serializer/buffered_file_writer.cpp	2
 common/serializer/binary_serializer.cpp	11
 common/serializer/memory_stream.cpp	5
-common/sort/comparators.cpp	72
+common/sort/comparators.cpp	103
```
[the remark this answers is not in the document]
Comment (Maxxen) [reply .github/config/uncovered_files.csv:61]: I spent a lot of time trying to figure out how to trigger these but couldn't work it out completely. Seeing as it's more or less the same as for normal lists It didn't seem to be that critical. But for sure something to look into in the future.

File: test/sql/types/nested/array/array_storage_2.test
[the remark this answers is not in the document]
Comment (Maxxen) [reply test/sql/types/nested/array/array_storage_2.test:1]: I don't really know what to call them, they just stress-test the storage in various ways :/

File: src/common/types.cpp [hunk 13]
```diff
@@ -704,6 +725,11 @@ LogicalType LogicalType::MaxLogicalType(const LogicalType &left, const LogicalTy
 		auto new_child = MaxLogicalType(ListType::GetChildType(left), ListType::GetChildType(right));
 		return LogicalType::LIST(new_child);
 	}
+	if (type_id == LogicalTypeId::ARRAY) {
+		auto new_child = MaxLogicalType(ArrayType::GetChildType(left), ArrayType::GetChildType(right));
+		auto new_size = MaxValue<uint32_t>(ArrayType::GetSize(left), ArrayType::GetSize(right));
```
Comment (Mytherin) [inline src/common/types.cpp:730]: `idx_t`, or maybe we should have a limit (100K)?

File: src/common/types/vector.cpp [hunk 14]
```diff
@@ -1201,6 +1336,53 @@ void Vector::Verify(Vector &vector_p, const SelectionVector &sel_p, idx_t count)
...
+		//	1.2 OR The child vector is a CONSTANT_VECTOR and must be NULL
+		//  1.3 OR The child vector is a CONSTANT_VECTOR and array_size = 1
+		// 2. if the array vector is a FLAT_VECTOR, the child vector is a FLAT_VECTOR
+		// 	2.2 the count of the child vector is array_size * (parent)count
+
+		auto &child = ArrayVector::GetEntry(*vector);
+		auto array_size = ArrayType::GetSize(type);
+
+		if (child.GetVectorType() == VectorType::CONSTANT_VECTOR) {
+			D_ASSERT(ConstantVector::IsNull(child) || array_size == 1);
```
Comment (Mytherin) [inline src/common/types/vector.cpp:1352]: array_size=1 probably doesn't have to be here

File: src/core_functions/function_list.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (Mytherin) [reply src/core_functions/function_list.cpp:20]: All functions except for the super core functions should go into `core_functions`. The idea is that the `core_functions` will eventually move to an extension, so you can have DuckDB without this set of functions (although in general all DuckDB installations will include these by default). This is not possible for all functions, for example `min` and `max` are required to implement other machinery (e.g. subqueries) so they must ALWAYS be bundled with DuckDB.

**Review by Mytherin — APPROVED** [review]
Looks great! Some minor comments - otherwise good to go from my side.

**Comment by Maxxen** [comment]
All green! @Mytherin

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**